server/src/main/java/keywhiz/service/daos/AclDAO.java

```diff
@@ -578,17 +578,24 @@ protected void evictClient(Configuration configuration, long clientId, long grou
     private final SecretSeriesDAOFactory secretSeriesDAOFactory;
     private final ClientMapper clientMapper;
     private final GroupMapper groupMapper;
-    private final SecretSeriesMapper secretSeriesMapper;
+    private final SecretSeriesMapper.SecretSeriesMapperFactory secretSeriesMapperFactory;
     private final SecretContentMapper secretContentMapper;
     private final RowHmacGenerator rowHmacGenerator;
     private final KeywhizConfig config;
 
-    @Inject public AclDAOFactory(DSLContext jooq, @Readonly DSLContext readonlyJooq, ClientDAOFactory clientDAOFactory,
-                                 GroupDAOFactory groupDAOFactory, SecretContentDAOFactory secretContentDAOFactory,
-                                 SecretSeriesDAOFactory secretSeriesDAOFactory, ClientMapper clientMapper,
-                                 GroupMapper groupMapper, SecretSeriesMapper secretSeriesMapper,
-                                 SecretContentMapper secretContentMapper, RowHmacGenerator rowHmacGenerator,
-                                 KeywhizConfig config) {
+    @Inject public AclDAOFactory(
+        DSLContext jooq,
+        @Readonly DSLContext readonlyJooq,
+        ClientDAOFactory clientDAOFactory,
+        GroupDAOFactory groupDAOFactory,
+        SecretContentDAOFactory secretContentDAOFactory,
+        SecretSeriesDAOFactory secretSeriesDAOFactory,
+        ClientMapper clientMapper,
+        GroupMapper groupMapper,
+        SecretSeriesMapper.SecretSeriesMapperFactory secretSeriesMapperFactory,
+        SecretContentMapper secretContentMapper,
+        RowHmacGenerator rowHmacGenerator,
+        KeywhizConfig config) {
```

Comment:
Do you mean in later in this PR? :)

Was there a specific reason or motivation for doing so? (asking as someone with no context on the factory pattern, so bear with me...)

Reply:
It all comes down to scoping. Before, the mapper didn't have any database dependencies so it didn't have to care about whether there were any transactions or connections currently in use. Now, it has to make a call into the groups DAO in order to fetch the name of the owning group (if any), which means that it needs to piggyback off of any existing connection/transaction in order to prevent resource leaks. So, every time we need an instance of the mapper we need to pass in the current database "scope", which is what the factory does for us.

There are fancier ways to accomplish this in Guice, either via auto-generated factories or child injectors (my preference). I was planning on putting together a PR with the latter today if I had time.

Reply:
Ah ok. That does make me wonder though, how did we get by without database dependencies before?

Reply:
The way it works is that it's using a Java Stream to read the raw records from the database, and the translation class is responsible for mapping that into our own model class. Since the translator was starting from the raw record it didn't have to perform any database operations of its own.

I have a nagging feeling that there's a more idiomatic way to accomplish what I'm trying to do (perhaps a subclass of the Jooq entity which includes a field that's populated via a join), and I'll be doing some more exploration of this later this week, but this works for now.